File: Data Structures & Algorithms/string-encode-and-decode/submission-2.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        s = ""
        for st in strs:
            s += str(len(st)) + ":" + st
        return s

    def decode(self, s: str) -> List[str]:
        res = []
        i = 0
        while i < len(s):
            j = i
            while s[j] != ':':
                j += 1
            length = int(s[i:j])
            x = s[j+1:j+length+1]
            res.append(x)
            i = j + length + 1
            # print(i, length)

        return res

            

File: Data Structures & Algorithms/anagram-groups/submission-2.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> res;
        unordered_map<string, vector<string>> umap;
        for(auto str: strs) {
            auto str_cpy = str;
            sort(str_cpy.begin(), str_cpy.end());
            umap[str_cpy].push_back(str);
        }

        for(auto elem: umap) {
            res.push_back(elem.second);
        }
        return res;
    }
};
